### PR TITLE
Fix S2222 FP and FN: Support Monitor.IsEntered

### LINKED ISSUE #5519: Fix S2222 FP and FN: Support Monitor.IsEntered
`Monitor.IsEntered` should define lock state for the argument object
```
        public void IsEntered_Compliant()
        {
            Monitor.Enter(obj);      // FP
            if (Monitor.IsEntered(obj))
            {
                Monitor.Exit(obj);
            }
        }

        public void IsEntered_Acquired()
        {
            if (Monitor.IsEntered(obj)) // FN
            {
                if (condition)
                {
                    Monitor.Exit(obj);
                }
            }
        }
```

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #5519

### CODE REVIEW

**Comment by pavel-mikula-sonarsource** [comment]
@csaba-sagi-sonarsource Shall we ship this to 8.37 in case we can't release today?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs [hunk 1]
```diff
@@ -107,17 +107,34 @@ public override ProgramState ConditionEvaluated(SymbolicContext context)
...
-                && property.Instance.TrackedSymbol() is { } lockSymbol)
+                && property.Property.ContainingType.IsAny(KnownType.System_Threading_ReaderWriterLock, KnownType.System_Threading_ReaderWriterLockSlim))
             {
-                return context.State[context.Operation].HasConstraint(BoolConstraint.True)  // Is it a branch with the Lock held?
-                    ? AddLock(context, lockSymbol)
-                    : RemoveLock(context, lockSymbol);
+                return ProcessCondition(property.Instance.TrackedSymbol());
+            }
+            else if (context.Operation.Instance.AsInvocation() is { } invocation
+                && invocation.TargetMethod.Is(KnownType.System_Threading_Monitor, "IsEntered"))
```
Comment (csaba-sagi-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs:115]: NITPICK
```suggestion
            else if (context.Operation.Instance.AsInvocation() is { } invocation
                       && invocation.TargetMethod.Is(KnownType.System_Threading_Monitor, "IsEntered"))
```

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!

**Comment by csaba-sagi-sonarsource** [comment]
> @csaba-sagi-sonarsource Shall we ship this to 8.37 in case we can't release today?

If we can release Today, I think we can go without this fix, just to make sure we release as soon as possible to be able to deliver the rule to the next SQ release. 
If, we cannot release today, I suggest merging the fix, and doing the re-validation on peachee.
Also, if we will not make it for the SQ release, we don't necessarily need to release 8.37 and right after 8.38 with the hardening. I would merge the two releases into one.

**Pull request merged**